2015-Fall/oPolo/Calculator - assignment 2/Calculator/CalculatorBrain.swift

```diff
@@ -0,0 +1,177 @@
+        case Variable(String)
+        
+        var description: String {
+            get {
+                switch self {
+                case .Operand(let operand):
+                    return "\(operand)"
+                case .Constant(let constant, _):
+                    return constant
+                case .UnaryOperation(let symbol, _):
+                    return symbol
+                case .BinaryOperation(let symbol, _):
+                    return symbol
+                case .Variable(let variable):
+                    return variable
+                }
+            }
+        }
+    }
+    
+    let calculationQueueChangedFunctionName = "QueuedCalculationsChanged"
+    private var opStack = [Op]() {
+        didSet {
+            //Fires a notification when a new operand/operator/constant/etc is added to the stack.
+            NSNotificationCenter.defaultCenter().postNotificationName(calculationQueueChangedFunctionName, object: nil)
+        }
+    }
+    
+    
+    private var knownOps = [String:Op]()
+    
+    init() {
+        func learnOp(op: Op) {
+            knownOps[op.description] = op
+        }
+        
+        learnOp(Op.Constant("π", M_PI))
+        learnOp(Op.BinaryOperation("×", *))
+        learnOp(Op.BinaryOperation("÷") { $1 / $0})
+        learnOp(Op.BinaryOperation("+", +))
+        learnOp(Op.BinaryOperation("−") { $1 - $0})
+        learnOp(Op.UnaryOperation("√", sqrt))
+        learnOp(Op.UnaryOperation("sin", sin))
+        learnOp(Op.UnaryOperation("cos", cos))
+    }
+    
+    private func evaluate(ops: [Op]) -> (result: Double?, remainingOps: [Op]) {
+        if !ops.isEmpty {
+            var remainingOps = ops
+            let op = remainingOps.removeLast()
+            switch op {
+            case .Operand(let operand):
+                return (operand, remainingOps)
+            case .Variable(let variable):
+                return (variableValues[variable], remainingOps)
+            case .Constant(_, let constantValue):
+                return (constantValue, remainingOps)
+            case .UnaryOperation(_, let operation):
+                let operandEvaluation = evaluate(remainingOps)
+                if let operand = operandEvaluation.result {
+                    return (operation(operand), operandEvaluation.remainingOps)
+                }
+            case .BinaryOperation(_, let operation):
+                let op1Evaluation = evaluate(remainingOps)
+                if let operand1 = op1Evaluation.result {
+                    let op2Evaluation = evaluate(op1Evaluation.remainingOps)
+                    if let operand2 = op2Evaluation.result {
+                        return (operation(operand1, operand2), op2Evaluation.remainingOps)
+                    }
+                }
+            }
+        }
+        return (nil, ops)
+    }
+    
+    
+    var description : String {
+        get {
+            var (returnedDescription, remainingOps) = createDescription(opStack)
```

Comment:
I'm not really following the logic with the while loop. Let's discuss it in real life :)